packages/core/src/test/notifications/controller.test.ts

```diff
@@ -0,0 +1,506 @@
+/*!
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import * as vscode from 'vscode'
+import assert from 'assert'
+import sinon from 'sinon'
+import { NotificationsController, NotificationsNode, RuleEngine } from '../../notifications'
+import * as HttpResourceFetcherModule from '../../shared/resourcefetcher/httpResourceFetcher'
+import globals from '../../shared/extensionGlobals'
+import { NotificationData, ToolkitNotification } from '../../notifications/types'
+import { _useLocalFilesCheck } from '../../notifications/controller'
+import { randomUUID } from '../../shared'
+import { installFakeClock } from '../testUtil'
+import * as FakeTimers from '@sinonjs/fake-timers'
+
+describe('Notifications Controller', function () {
+    const panelNode: NotificationsNode = new NotificationsNode('toolkit')
+    let controller: NotificationsController
+    const ruleEngine: RuleEngine = new RuleEngine({
+        ideVersion: '1.83.0',
+        extensionVersion: '1.20.0',
+        os: 'LINUX',
+        computeEnv: 'local',
+        authTypes: ['builderId'],
+        authRegions: ['us-east-1'],
+        authStates: ['connected'],
+        authScopes: ['codewhisperer:completions', 'codewhisperer:analysis'],
+        installedExtensions: ['ext1', 'ext2', 'ext3'],
+        activeExtensions: ['ext1', 'ext2'],
+    })
+
+    let clock: FakeTimers.InstalledClock
+    let sandbox: sinon.SinonSandbox
+    let getNewETagContentStub: sinon.SinonStub
+    let ruleEngineSpy: sinon.SinonSpy
+    let focusPanelSpy: sinon.SinonSpy
+    let fetchUrl: string
+
+    function dismissNotification(notification: ToolkitNotification) {
+        // We could call `controller.dismissNotification()`, but this emulates a call
+        // from actually clicking it in the view panel.
+        return vscode.commands.executeCommand('_aws.toolkit.notifications.dismiss', {
+            getTreeItem: () => {
+                const item = new vscode.TreeItem('test')
+                Object.assign(item, {
+                    command: { arguments: [notification] },
+                })
+
+                return item
+            },
+        })
+    }
+
+    before(function () {
+        assert.ok(!_useLocalFilesCheck)
+        clock = installFakeClock()
+    })
+
+    beforeEach(async function () {
+        panelNode.setNotifications([], [])
+        controller = new NotificationsController('toolkit', panelNode)
+
+        sandbox = sinon.createSandbox()
+
+        getNewETagContentStub = sandbox.stub()
+        ruleEngineSpy = sandbox.spy(ruleEngine, 'shouldDisplayNotification')
+        focusPanelSpy = sandbox.spy(panelNode, 'focusPanel')
+        sandbox.stub(HttpResourceFetcherModule, 'HttpResourceFetcher').callsFake(function (this: any, url: string) {
```

Comment:
The tests in this PR do a nice job of minimizing mocks/stubs. Does it help at all to plumb this "fetcher" abstraction through as a parameter (to e.g. the constructor)? After all, I think the Notifcations has a singleton which presumably is intended to serve the purpose of injecting stuff.